test/test_allocation_domain.cpp

```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
       __FILE__);
 }
 
+// A global->global copy kernel where both inputs are NHWC memory format
+TEST_F(AllocationDomainTest, NHWC4d_To_NHWC4d_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(4);
+  fusion.addInput(tv0);
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  std::vector<IterDomain*> tv0_nhwc = {
+      tv0->axis(0), tv0->axis(2), tv0->axis(3), tv0->axis(1)};
+  tv0->setAllocationDomain(tv0_nhwc, true);
+
+  std::vector<IterDomain*> tv1_nhwc = {
+      tv1->axis(0), tv1->axis(2), tv1->axis(3), tv1->axis(1)};
+  tv1->setAllocationDomain(tv1_nhwc, true);
+
+  // [N, C, H, W]
+  tv1->reorder({{1, -1}});
+  // [N, H, W, C]
+  tv1->merge(0);
+  tv1->merge(0);
+  tv1->merge(0);
+  // [N*H*W*C]
+  tv1->split(0, 4);
+  tv1->axis(1)->parallelize(ParallelType::Vectorize);
+  tv1->split(0, 128);
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+  tv1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, TIDx, V]
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  int n = 31, h = 64, w = 103, c = 21;
+
+  at::Tensor t0_wrong_format = at::randn({n, c, h, w}, options);
+  at::Tensor t0 = t0_wrong_format.contiguous(at::MemoryFormat::ChannelsLast);
```

Comment:
nitpick. maybe it's me being paranoid. But I'm pretty uncomfortable with pytorch's implicit memory_format conversion. Can we use explicit calls like `as_strided` instead? so it's easier to read the examples.

Reply:
Why `as_strided`? If you don't like `contiguous`, then isn't the alternative `permute` instead of `as_strided`?

Reply:
lol, not arguing what's the proper way to construct inputs here. Feel free to go with what you think fits here. (even `contiguous` is probably fine, as the API is somewhat stable, but might catch you by surprise when you have broadcast dimensions - i.e. size-1 - in inputs).

I guess `as_strided` feels a lot more straightforward to me. i.e. you are providing input sizes along with their strides, so there's nothing like `permute` which actually moves dimensions around and changes their semantic order. :man_shrugging:

Reply:
Wait, no, I think neither `contiguous` nor `permute` works. It has to be `contiguous`. Because I want to keep the semantics of the tensor, but store it differently. Neither `as_strided` nor `permute` change the underlying tensor storage.

Reply:
> Wait, no, I think neither `as_strided`  nor `permute` works. It has to be `contiguous`.
^^^ I don't agree with this.
I agree that `contiguous` is different from the other two and they can't be used interchangeably. But in our test, we don't care about the exact data, since we control the reference as well.

i.e. we can create a contiguous buffer and call `as_strided` to create a tensor with a semantic size and an arbitrary memory storage. As long as: 1. it matches fusion definition; 2. the same tensor is used for reference computation.

Reply:
similarly, in order to create a [2, 3, 4, 5] tensor in channels-last format, you can create a [2, 4, 5, 3] tensor an permute it to get a size [2, 3, 4, 5] in channels-last format.
It's not going to be the same tensor, but it doesn't matter for our tests.

Reply:
I updated the test to use `as_strided`. In terms of interface, I definitely prefer the `contiguous(at::MemoryFormat::ChannelsLast)` one. It is very unfortunate that this interface is used for an unpredictable and confusing behavior.